File: frontend/src/services/multas.ts
import api from './api'
import type { Multa, MultaListItem } from '@/types'

interface MultaFilters {
  q?: string
  status?: string
  condutor_id?: number
  skip?: number
  limit?: number
}

export const multaService = {
  async listar(filters: MultaFilters = {}): Promise<MultaListItem[]> {
    const response = await api.get<MultaListItem[]>('/multas', { params: filters })
    return response.data
  },

  async detalhe(id: number): Promise<Multa> {
    const response = await api.get<Multa>(`/multas/${id}`)
    return response.data
  },

  async criar(data: Partial<Multa>): Promise<Multa> {
    const response = await api.post<Multa>('/multas', data)
    return response.data
  },

  async atualizar(id: number, data: Partial<Multa>): Promise<Multa> {
    const response = await api.patch<Multa>(`/multas/${id}`, data)
    return response.data
  },

  async excluir(id: number): Promise<void> {
    await api.delete(`/multas/${id}`)
  },
}

File: backend/app/api/v1/endpoints/multas.py
from fastapi import APIRouter, Depends, Query

from app.core.dependencies import DatabaseDep, get_current_user, require_perfil
from app.repositories.multas import MultaRepository, TipoMultaRepository
from app.schemas.common import MessageResponse
from app.schemas.multas import MultaCreate, MultaListItem, MultaResponse, MultaUpdate, TipoMultaResponse
from app.services.multas import MultaService

router = APIRouter(prefix="/multas", tags=["Multas"])


@router.get("/tipos", response_model=list[TipoMultaResponse], dependencies=[Depends(get_current_user)])
async def listar_tipos_multa(db: DatabaseDep):
    return await TipoMultaRepository(db).get_ativos()


@router.get("", response_model=list[MultaListItem], dependencies=[Depends(get_current_user)])
async def listar_multas(
    db: DatabaseDep,
    q: str = Query("", description="Busca por placa"),
    status: str = Query(""),
    condutor_id: int | None = Query(None),
    skip: int = 0,
    limit: int = 100,
):
    return await MultaRepository(db).search(q=q, status=status, condutor_id=condutor_id, skip=skip, limit=limit)


@router.post("", response_model=MultaResponse, dependencies=[Depends(require_perfil("ADMIN", "GESTOR"))])
async def criar_multa(data: MultaCreate, db: DatabaseDep):
    service = MultaService(MultaRepository(db))
    return await service.criar(data)


@router.get("/{multa_id}", response_model=MultaResponse, dependencies=[Depends(get_current_user)])
async def detalhe_multa(multa_id: int, db: DatabaseDep):
    service = MultaService(MultaRepository(db))
    return await service.get_or_404(multa_id)


@router.patch("/{multa_id}", response_model=MultaResponse, dependencies=[Depends(require_perfil("ADMIN", "GESTOR"))])
async def atualizar_multa(multa_id: int, data: MultaUpdate, db: DatabaseDep):
    service = MultaService(MultaRepository(db))
    return await service.atualizar(multa_id, data)


@router.delete("/{multa_id}", response_model=MessageResponse, dependencies=[Depends(require_perfil("ADMIN"))])
async def excluir_multa(multa_id: int, db: DatabaseDep):
    service = MultaService(MultaRepository(db))
    multa = await service.get_or_404(multa_id)
    await MultaRepository(db).delete(multa)
    return MessageResponse(message="Multa excluída com sucesso")
